### DISCUSSION TITLE
Export of Langium AST types

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1197] [Q&A] [answered]
Hi, is there a reason the only exported AST type is `Grammar` https://github.com/eclipse-langium/langium/blob/f3cf20687c8a891935bb3425f9d12242aeae37ab/packages/langium/src/index.ts#L26

(in langium 1x we can do somthing like`import type { AbstractElement, CrossReference, ParserRule } from "langium/lib/grammar/generated/ast"`)

**Comment by msujew** [comment, marked as answer]
@cdietrich To prevent pollution of the imports, all other internal ast types/methods are grouped in the `GrammarAST` object. You can access them from there.

> (in langium 1x we can do somthing like `import type { AbstractElement, CrossReference, ParserRule } from "langium/lib/grammar/generated/ast"`)

This was never recommended, but it's not possible in commonJS to prevent adopters from arbitrarily importing stuff from packages. ESM now enforces our exports.

**Reply by cdietrich** [reply]
is 
```
port type { GrammarAST as grammar } from "langium".


x: grammar.AbstractElement
```
also just acc. working?

**Reply by msujew** [reply]
No, it's completely fine to rename imports, that's a normal ESM feature.
